### ISSUE TITLE
Reduce RAM utilization

### ISSUE THREAD

**Issue by jon12156** [panel-attack/panel-game#21]
Perhaps we can NOT load all assets and sounds at startup.  Just load before a round the assets and sounds that will be needed.

We'll need to be sure starting of 2p_net_vs rounds is not negatively affected if it takes opponents different amounts of time to do this loading of assets and sounds.

**Comment by JamesVanBoxtel** [comment]
Has anyone complained about memory usage? Doesn't seem too bad on my system, maybe on mobile or people that have massive character sets?

**Comment by Endaris** [comment]
I think loading once at the start to make sure the game knows which mods are actually usable makes many things around mods easier to account for.

Including a configuration for fallback characters/stages that get used if you don't have the opponent's mods would help a lot with preventing the loading of many characters and stages over the course of a set in the first place as the majority of people don't even switch their character in the middle of a set.

In general it might be worthwhile to somehow try and track the RAM usage of PA because I think that it typically slowly increases over the course of a session.

**Comment by Endaris** [comment]
So when I did the character SFX thing I realised there's actually not a whole lot to do here.
Basically the game keeps in memory:
Your entire theme, no brainer, needs to be in memory.
The entire character in your config; not strictly necessary at all times (e.g. not while spectating) but makes a lot of sense
Of all characters: selection SFX and icon for select_screen; not strictly necessary at all times (e.g. Time Attack) but makes a lot of sense
Of all stages: The thumbnail pic for display in select_screen; not strictly necessary at all times (e.g. Time Attack) but makes a lot of sense

While in a select_screen game mode: All characters and stages that have been picked. These are maintained until you quit out to net vs lobby or main menu where they'll all be cleared minus the character/stage you had last selected and minus the character/stage your opponent last selected.
We could attempt to clear out stages/characters when they are unpicked via a non-random selection. And we could generally include the opponent assets in the clear if coming out from online vs. These are likely going to be the biggest angles for keeping RAM low as the loading of normal/danger music assets will add up the most.

All panel sets are always fully loaded.
We could cut down on that by only loading the images required for display in select_screen + the entire set selected in the config but it's honestly not going to do a whole lot memory wise, even if you have the entire HD pack installed as one panel set is typically not more than 100-150. Might be worthwhile once UHD panels become more popular as they're to be expected at 3-4x that size (6-8x for 4k) but not a thing atm.

The one spot where RAM truly goes ham is sound test because PA will just straight up load your entire music and SFX into memory. And yeah, that will collapse your game if you have massive character sets. But again, these will all get unloaded when going back to main menu and we're tracking that as a separate issue in panel-attack/panel-game#118 so I think that shouldn't have relevance here.

Overall conclusion:
One thing that would contribute massively would be to add a client-side configuration for automatically picking the same character for the opponent every time if their character is not available. Most people don't switch characters/stages a lot within a set and the only reason why loading keeps being done is that one doesn't have their character or the server selected stage and random rolls for new character/stage music every game again.
Similarly, implementing panel-attack/panel-game#148 would allow us to implement a check against that option to prevent excessive asset loading in the respective `sound_init` functions.
